### PR TITLE
state: getBlockByNumber

### LINKED ISSUE #460: GetBlockByNumber
Current;ly no way to lookup block by number. Until we have an immediate need we should leave this open.

### PR DESCRIPTION
**Pull request by dziabko**
Added mapping of blockNumber to blockHash. Implemented getBlockByNumber using the mapping & getBlockByHash.

## Changes
- Added mapping of block number to block hash in the DB
- Used mapping & getBlockByHash to implement getBlockByNumber

## Tests:
```

```

### Issues:
closes #460

### CODE REVIEW

File: state/block.go [hunk 1]
```diff
@@ -24,15 +25,28 @@ func NewBlockState() *blockState {
...
-	headerPrefix    = []byte("hdr") // headerPrefix + hash -> header
-	blockDataPrefix = []byte("hsh") // blockDataPrefix + hash -> blockData
+	headerPrefix     = []byte("hdr") // headerPrefix + hash -> header
+	blockDataPrefix  = []byte("bld") // blockDataPrefix + hash -> blockData
+	headerHashPrefix = []byte("hsh")
 )
 
+// encodeBlockNumber encodes a block number as big endian uint64
+func encodeBlockNumber(number uint64) []byte {
+	enc := make([]byte, 8)
```
Comment (ryanchristo) [inline state/block.go]: Maybe a comment about why we are using `8` here.

File: state/block.go [hunk 2]
```diff
@@ -24,15 +25,28 @@ func NewBlockState() *blockState {
 
 var (
 	// Data prefixes
-	headerPrefix    = []byte("hdr") // headerPrefix + hash -> header
-	blockDataPrefix = []byte("hsh") // blockDataPrefix + hash -> blockData
+	headerPrefix     = []byte("hdr") // headerPrefix + hash -> header
+	blockDataPrefix  = []byte("bld") // blockDataPrefix + hash -> blockData
+	headerHashPrefix = []byte("hsh")
```
Comment (ryanchristo) [inline state/block.go]: Maybe a comment for `headerHashPrefix`.

File: state/block.go (hunk 1 above)
Comment (noot) [reply state/block.go]: it's because in the line after we are doing PutUint64, which takes 8 bytes to put the uint64 into

**Comment by ryanchristo** [comment]
Let's add a test before merging.

**Comment by ryanchristo** [comment]
cc @dziabko can you please add tests for the following after updating the branch and then include a test for each state method added in future pull requests.

- `AddBlock`
- `GetBlockByHash`
- `GetBlockByNumber`

File: state/block_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/ChainSafe/gossamer/common"
+	"github.com/ChainSafe/gossamer/core/types"
+)
+
+func TestGetBlockByNumber(t *testing.T) {
+	dataDir := "../test_data/block"
```
Comment (ansermino) [inline state/block_test.go]: Please clean up any test artifacts. Check out the top `cmd/gossamer/configcmd_test.go` for some copy pasta

File: state/block.go [hunk 4]
```diff
@@ -12,8 +12,8 @@ import (
 )
 
 type blockState struct {
-	bt          *blocktree.BlockTree
-	db          *polkadb.BlockDB
+	Bt          *blocktree.BlockTree
+	Db          *polkadb.BlockDB
```
Comment (ansermino) [inline state/block.go]: Do these need to be exported?

File: state/block_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		t.Fatal(err)
+	}
+	//Close DB & erase data dir contents
+	defer func() {
+		err = blockState.Db.Db.Close()
+		if err != nil {
+			t.Fatal("BlockDB close err: ", err)
+		}
+		if err = os.RemoveAll(dataDir); err != nil {
+			fmt.Println("removal of temp directory test_data failed")
```
Comment (ryanchristo) [inline state/block_test.go]: should we `t.Fatal(err)` here rather than print?

File: state/block_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		}
+	}()
+
+	// Create a header & blockdata
+	blockHash := common.NewHash([]byte{0, 1, 2})
+	header := types.BlockHeaderWithHash{
```
Comment (ryanchristo) [inline state/block_test.go]: should we name `blockHeader` to be consistent?

File: state/block_test.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	dataDir := "../test_data/block"
+
+	//Create a new blockState
+	blockState, err := NewBlockState(dataDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+	//Close DB & erase data dir contents
+	defer func() {
+		err = blockState.Db.Db.Close()
```
Comment (ryanchristo) [inline state/block_test.go]: Should we create a `Stop` method on `blockState` that includes `Db.Db.Close` and `os.RemoveAll` to clean this up and also to make it reusable in multiple tests?

**Comment by ryanchristo** [comment]
cc @dziabko What's the status of this pull request? Do you have any questions or need any help?

**Comment by ryanchristo** [comment]
cc @dziabko what's the status of this pull request?

File: state/block_test.go (hunk 3 above)
Comment (thomasmodeneis) [reply state/block_test.go]: this same test is in the #503 , I've added comments there with snippets for the fix, could this be applied there instead ?

File: state/block_test.go (hunk 7 above)
Comment (thomasmodeneis) [reply state/block_test.go]: this same test is in the #503 , I've added comments there with snippets for the fix, could this be applied there instead ?

File: state/block_test.go (hunk 5 above)
Comment (thomasmodeneis) [reply state/block_test.go]: this could be removed completely and use ioutils intead, see

File: state/block_test.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	dataDir := "../test_data/block"
+
+	// Create & start a new State service
+	stateService := NewService(dataDir)
+	stateService.Start()
+
+	// Close the service, and remove dataDir once test is done
+	defer stateService.Stop()
+	defer func() {
+		if err := os.RemoveAll("../test_data"); err != nil {
```
Comment (ryanchristo) [inline state/block_test.go]: Should we use the `dataDir` variable for this? ie, `os.RemoveAll(dataDir)`

File: state/block_test.go (hunk 3 above)
Comment (ryanchristo) [reply state/block_test.go]: It would be nice to apply here but I'm ok with it being addressed in #503.

File: state/block.go [hunk 9]
```diff
@@ -102,19 +115,28 @@ func (bs *blockState) GetLatestBlock() types.BlockHeader {
...
+func (bs *blockState) GetBlockByNumber(n *big.Int) (types.Block, error) {
+	// First retrieve the block hash based on the block number from the database
+	hash, err := bs.db.Db.Get(headerHashKey(n.Uint64()))
+
+	if err != nil {
+		return types.Block{}, err
+	}
+
+	// Then find the block based on the hash
+	endHash := common.NewHash(hash)
```
Comment (ryanchristo) [inline state/block.go]: Is `endHash` a typo?

Comment (ryanchristo) [reply state/block.go]: Maybe rename this `hash`, then rename the current `hash` to `data` or `bytes`.

File: state/block_test.go (hunk 5 above)
Comment (ryanchristo) [reply state/block_test.go]: updated to `t.Fatal(err)`, let's address the latter in #503

File: state/block_test.go (hunk 6 above)
Comment (ryanchristo) [reply state/block_test.go]: looks like this has been addressed

File: state/block_test.go (hunk 7 above)
Comment (ryanchristo) [reply state/block_test.go]: this has not been addressed but let's add it in #503 when we add multiple tests

File: state/block_test.go (hunk 3 above)
Comment (thomasmodeneis) [reply state/block_test.go]: done

File: state/block_test.go (hunk 8 above)
Comment (thomasmodeneis) [reply state/block_test.go]: done

File: state/block.go (hunk 9 above)
Comment (thomasmodeneis) [reply state/block.go]: done

**Comment by thomasmodeneis** [comment]
PR comments accepted, code updated, ready to merge.

**Pull request merged**